rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+            };
+
+            self.comment = unsafe { (*(self.comment)).node.next.cast::<yp_comment_t>() };
+            Some(current)
+        }
+    }
+}
+
+/// The result of parsing a source string.
+pub struct ParseResult<'pr> {
+    source: &'pr [u8],
+    parser: NonNull<yp_parser_t>,
+    node: NonNull<yp_node_t>
+}
+
+impl<'pr> ParseResult<'pr> {
+    /// Returns the source string that was parsed.
+    #[must_use]
+    pub const fn source(&self) -> &'pr [u8] {
+        self.source
+    }
+
+    /// Returns a slice of the source string that was parsed using the given
+    /// location range.
+    #[must_use]
+    pub fn slice(&self, location: &Location<'pr>) -> &'pr [u8] {
+        let ptr = self.source.as_ptr() as usize;
+        &self.source[(location.start() - ptr)..(location.end() - ptr)]
+    }
+
+    /// Returns an iterator that can be used to iterate over the errors in the
+    /// parse result.
+    #[must_use]
+    pub fn errors(&self) -> Diagnostics<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).error_list;
+            Diagnostics { diagnostic: list.head.cast::<yp_diagnostic_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the warnings in the
+    /// parse result.
+    #[must_use]
+    pub fn warnings(&self) -> Diagnostics<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).warning_list;
+            Diagnostics { diagnostic: list.head.cast::<yp_diagnostic_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns an iterator that can be used to iterate over the comments in the
+    /// parse result.
+    #[must_use]
+    pub fn comments(&self) -> Comments<'_> {
+        unsafe {
+            let list = &mut (*self.parser.as_ptr()).comment_list;
+            Comments { comment: list.head.cast::<yp_comment_t>(), marker: PhantomData }
+        }
+    }
+
+    /// Returns the root node of the parse result.
+    #[must_use]
+    pub fn node(&self) -> Node<'pr> {
+        Node::new(self.node.as_ptr())
+    }
+}
+
+impl<'pr> Drop for ParseResult<'pr> {
+    fn drop(&mut self) {
+        unsafe {
+            yp_node_destroy(self.parser.as_ptr(), self.node.as_ptr());
+
+            // TODO: Why isn't this working?
+            // yp_parser_free(self.parser.as_ptr());
+        }
+    }
+}
+
+/// Parses the given source string and returns a parse result.
+/// 
+/// # Panics
+///
+/// Panics if the parser fails to initialize.
+///
+#[must_use]
+pub fn parse(source: &[u8]) -> ParseResult<'_> {
+    unsafe {
+        let mut uninit = MaybeUninit::<yp_parser_t>::uninit();
+
+        yp_parser_init(
+            uninit.as_mut_ptr(),
+            source.as_ptr().cast::<i8>(),
+            source.len(),
+            std::ptr::null(),
+        );
+
+        let parser = NonNull::new_unchecked(uninit.assume_init_mut());
+        let node = NonNull::new_unchecked(yp_parse(parser.as_ptr()));
```

Comment:
for debugging purposes, I imagine it will be slightly more friendly to the debugger if the `yp_parse` invocation were on its own line.